addons/purchase_stock/models/purchase.py

```diff
@@ -99,8 +99,8 @@ def button_cancel(self):
             if order.state in ('draft', 'sent', 'to approve', 'purchase'):
                 for order_line in order.order_line:
                     order_line.move_ids._action_cancel()
-                    if order_line.move_dest_ids:
```

Comment:
maybe it's a better idea to directly add the condition:
```
if move.state in ('done', 'cancel'):
    continue
```
at the start of _recompute_state in stock/models/stock_move.py

cc: @sle-odoo